paddle/cinn/hlir/framework/group_scheduler.cc

```diff
@@ -0,0 +1,1224 @@
+      if (visited_var_names.count(name) > 0) {
+        continue;
+      }
+      visited_var_names.insert(name);
+      CHECK(for_map.find(block_name) != for_map.end());
+      CHECK(for_map[block_name].find(name) != for_map[block_name].end());
+      ir::Expr for_expr = for_map[block_name][name];
+      for (int i = 0; i < 3; ++i) {
+        if (for_type == for_expr.As<ir::For>()->for_type() &&
+            for_expr.As<ir::For>()->bind_info().offset == i &&
+            for_expr.As<ir::For>()->extent.get_constant() > 1) {
+          optim::ReplaceVarWithExpr(
+              &(indice_copies[i]), var.as_var_ref(), ir::Expr(1));
+          coef_and_ranges[i].second.min =
+              for_expr.As<ir::For>()->min.get_constant();
+          coef_and_ranges[i].second.max =
+              for_expr.As<ir::For>()->min.get_constant() +
+              for_expr.As<ir::For>()->extent.get_constant();
+        } else {
+          optim::ReplaceVarWithExpr(
+              &(indice_copies[i]), var.as_var_ref(), ir::Expr(0));
+        }
+      }
+    }
+    for (int i = 0; i < 3; ++i) {
+      VLOG(6) << "before simplify [" << i << "], the coefficient of "
+              << indice_value << " = " << indice_copies[i] << ", range = ("
+              << coef_and_ranges[i].second.min << ", "
+              << coef_and_ranges[i].second.max << ")";
+      indice_copies[i] = common::AutoSimplify(indice_copies[i]);
+      VLOG(6) << "after simplify [" << i << "], the coefficient of "
+              << indice_value << " = " << indice_copies << ", range = ("
+              << coef_and_ranges[i].second.min << ", "
+              << coef_and_ranges[i].second.max << ")";
+      coef_and_ranges[i].first =
+          static_cast<int>(indice_copies[i].get_constant());
+    }
+    return coef_and_ranges;
+  };
+
+  // Determine whether the indice of a pair of Store and Load cross CUDA threads
+  auto IsCrossThread = [&](ir::Expr store_indice_value,
+                           ir::Expr load_indice_value,
+                           const std::string& store_block_name,
+                           const std::string& load_block_name) {
+    Range store_thread_overall_range = CalculateRange(
+        store_indice_value, CudaBindInfo::kCudaThread, store_block_name);
+    Range load_thread_overall_range = CalculateRange(
+        load_indice_value, CudaBindInfo::kCudaThread, load_block_name);
+    Range store_serial_overall_range = CalculateRange(
+        store_indice_value, CudaBindInfo::kSerial, store_block_name);
+    Range load_serial_overall_range = CalculateRange(
+        load_indice_value, CudaBindInfo::kSerial, load_block_name);
+    auto store_thread_coefficient_and_range = GetCoefficientAndRange(
+        store_indice_value, ir::ForType::GPUThread, store_block_name);
+    auto load_thread_coefficient_and_range = GetCoefficientAndRange(
+        load_indice_value, ir::ForType::GPUThread, load_block_name);
+    VLOG(6) << "store_indice_value: " << store_indice_value;
```

Comment:
Can you just have some operator `<<` on `range`, so you don't have to handle log stream of range.min, range.max

Reply:
done